pyogrio/_io.pyx

```diff
@@ -1142,9 +1186,26 @@ def ogr_read(
         raise ValueError("'max_features' must be >= 0")
 
     try:
+        if encoding:
+            override_shape_encoding = True
+
+            # for shapefiles, SHAPE_ENCODING must be set before opening the file
+            # to prevent automatic decoding to UTF-8 by GDAL, so we save previous
+            # SHAPE_ENCODING so that it can be restored later
+            # (we do this for all data sources where encoding is set because
+            # we don't know the driver until after it is opened, which is too late)
+            encoding_b = encoding.encode("UTF-8")
+            encoding_c = encoding_b
```

Comment:
I originally wanted to keep `override_threadlocal_config_option` so that it only worked with C types, in case we ever wanted to pass in C values equivalent to what you'd get back from the function.  Otherwise, I think we'd want to have it encode both the key and value, even if the key passed in is a string defined in code (already a `const char*` I think?) rather than runtime.

Reply:
> even if the key passed in is a string defined in code (already a const char* I think?)

Hmm, not entirely sure that is correct, [cython docs](https://cython.readthedocs.io/en/latest/src/tutorial/strings.html#string-literals) say those are `str` (unicode) instances but yet show examples like this:  `cdef char* hello_world = 'hello world'` which implies that we at least get the conversion without extra overhead when defining string literals.